autogen/extensions/agent_optimizer.py

```diff
@@ -0,0 +1,440 @@
+}}
+(4) Directly return "TERMINATE":
+If you think there is no need to perform any other actions for the current function list since the current list is optimal more actions will harm the performance in future tasks. Please directly reply to me with "TERMINATE".
+
+One function signature includes the following five elements:
+1. Function name
+2. Function description
+3. JSON schema of arguments encoded as a string
+4. A list of package names imported by the function packages
+5. The code implementation
+
+Below are the signatures of the current functions:
+List A: {best_functions}.
+The following list are the function signatures that you have after taking {actions_num} actions to manipulate List A:
+List B: {incumbent_functions}.
+
+{accumerated_experience}
+
+Here are {best_conversations_num} conversation histories of solving {best_conversations_num} tasks using List A.
+History:
+{best_conversations_history}
+
+{statistic_informations}
+
+According to the information I provide, please take one of four actions to manipulate list B using the functions you know.
+Instead of returning TERMINATE directly or taking no action, you should try your best to optimize the function list. Only take no action if you really think the current list is optimal, as more actions will harm performance in future tasks.
+Even adding a general function that can substitute the assistant’s repeated suggestions of Python code with the same functionality could also be helpful.
+"""
+
+
+def execute_func(name, packages, code, **args):
+    """
+    The wrapper for generated functions.
+    """
+    pip_install = (
+        f"""print("Installing package: {packages}")\nsubprocess.run(["pip", "-qq", "install", "{packages}"])"""
+        if packages
+        else ""
+    )
+    str = f"""
+import subprocess
+{pip_install}
+print("Result of {name} function execution:")
+{code}
+args={args}
+result={name}(**args)
+if result is not None: print(result)
+"""
+    print(f"execute_code:\n{str}")
+    result = execute_code(str, use_docker="shaokun529/evoagent:v1")
+    if result[0] != 0:
+        raise Exception("Error in executing function:" + result[1])
+    print(f"Result: {result[1]}")
+    return result[1]
+
+
+class AgentOptimizer:
+    """
+    Base class for optimizing AutoGen agents. Specifically, it is used to optimize the functions used in the agent.
+    More information could be found in the following paper: https://arxiv.org/abs/2402.11359.
+    """
+
+    def __init__(
+        self,
+        max_actions_per_step: int,
+        config_file_or_env: Optional[str] = "OAI_CONFIG_LIST",
+        config_file_location: Optional[str] = "",
+        optimizer_model="gpt-4-1106",
```

Comment:
It is more clear to have optimizer_model explicitly specified during the call like max_actions_per_step instead of having default value.

Reply:
resolved